### PR TITLE
logger no longer returns an error

### PR DESCRIPTION
**Pull request by melekes**
See #50

### CODE REVIEW

**Comment by zramsay** [comment]
this works when pinned in glide.lock on tendermint. I also ran `make test` and it passed.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,18 +1,24 @@
 # Changelog
 
+## 0.4.0 (TBA)
+
+BREAKING CHANGES:
+
+- [log] logger functions no longer returns an error
```
Comment (ebuchman) [inline CHANGELOG.md]: let's just add this to 0.3.0

Comment (ebuchman) [reply CHANGELOG.md]: i got it :)

**Comment by ebuchman** [comment]
merged

**Comment by odeke-em** [comment]
Oops, I am too late to the wedding and passed the with the "hold your silence forever" threshold.

I realize that our goal is to pass the linter and that we never use the previously returned errors in our code. However, aren't we losing the functionality of getting back errors e.g when the logger is a distributed logger, for example using bugsnag or uber/go-zap or does tracing and tries to post the data, and the HTTP call fails? Or is this a conversation we'd rather have later on when we need that functionality?

**Comment by ebuchman** [comment]
That's a good conversation to have. But are we really going to check logging errors everywhere, or what ?

**Comment by melekes** [comment]
> when the logger is a distributed logger

@odeke-em and when it is gonna happen?

**Comment by odeke-em** [comment]
Good questions y'all.
`But are we really going to check logging errors everywhere, or what ?`
@ebuchman not really, and I mentioned that we aren't checking for any errors. However, some signatures like https://golang.org/pkg/fmt/#Printf return errors but no one every checks them, doesn't mean we should discard them entirely. On the other hand, I realize that we are being pragmatic here, so in deed if the need for catching errors then we'll have the conversation.

```
@odeke-em and when it is gonna happen?
```
@melekes I am not sure when, I was giving an example, and I assume that given that we are launching Cosmos in December 2017, perhaps people might want to have logs churned out say to Google Cloud logging or Stackdriver. However, no one is doing that right now though.

**Comment by melekes** [comment]
Good example with `fmt` package. Ok, so let's suppose somebody does request a possibility to log to Stackdriver. Will we be forced to go through all of our code and add error checking for all `Log` calls then?

**Comment by odeke-em** [comment]
So it seems to me like if they need to use a tmlibs.Logger, they'll still need to conform to the new signatures but can entirely make their own wrapper and do the error handling on their own (please note that these aren't fully thought out examples)
```go
type logToStackDriver struct {
   stLogger *stackdriver.Logger
   errHandler ErrHandler
}

var _ tmlibs.Logger = (*logToStackDriver)(nil)
func (ltsd *logToStackDriver) Debug(msg string, keyvals ...interface{}) {
    if err := ltsd.stLogger.Debug(msg, keyvals...); err != nil {
       ltsd.errHandler("debug", err)
    }
} 

func (ltsd *logToStackDriver) Info(msg string, keyvals ...interface{}) {
    if err := ltsd.stLogger.Info(msg, keyvals...); err != nil {
       ltsd.errHandler("info", err)
    }
} 

func (ltsd *logToStackDriver) Error(msg string, keyvals ...interface{}) {
    if err := ltsd.stLogger.Error(msg, keyvals...); err != nil {
       ltsd.errHandler("error", err)
    }
} 
```

**Comment by melekes** [comment]
They can write their own driver, for sure. But how do they enable it (replace current logger)?? By forking tendermint?

**Comment by odeke-em** [comment]
@melekes fortunately we made tmlibs.Logger an interface instead of a concrete type, so notice in the code snippet that I put up in https://github.com/tendermint/tmlibs/pull/51#issuecomment-330953559, that logger can now be used in places where tmlibs.Logger is required. I hope I am answering your question.

**Comment by melekes** [comment]
OK. here is the thing: you are talking about people using `log` package and building new drivers on top of it. But it never was my intent to build log package in such a way. log package was created exclusively for Tendermint. If people will want to forward their logs to some external storage (say Logstash), I would suggest to them to redirect Tendermint output to some log forwarder (i.e. `tendermint node > XXXX`. Does it make sense to you?

**Pull request closed without merging**